cassiopeia/data/CassiopeiaTree.py

```diff
@@ -0,0 +1,357 @@
+            self.__network = tree
+        if isinstance(tree, str):
+            self.__network = utilities.newick_to_networkx(tree)
+        if isinstance(tree, ete3.Tree):
+            self.__network = utilities.ete3_to_networkx(tree)
+
+        if self.__network is not None:
+
+            # add character states
+            for n in self.nodes:
+                if n in self.character_matrix.index.values:
+                    self.__network.nodes[n]['character_states'] = self.character_matrix.loc[n].to_list()
+                else:
+                    self.__network.nodes[n]['character_states'] = []
+            
+            # instantiate branch lengths
+            for u,v in self.edges:
+                self.__network[u][v]['length'] = 1
+
+            # instantiate node ages and edge depth
+            self.__network.nodes[self.root]['age'] = 0
+            self.__network.nodes[self.root]['edge_depth'] = 0
+            for u,v in self.depth_first_traverse_edges(source=self.root):
+                self.__network.nodes[v]['age'] = self.__network.nodes[u]['age'] + self.__network[u][v]['length']
+                self.__network.nodes[v]['edge_depth'] = self.__network.nodes[v]['age']
+
+    @property
+    def n_cell(self) -> int:
+        """Returns number of cells in character matrix.
+        """
+        return self.character_matrix.shape[0]
+
+    @property
+    def n_character(self) -> int:
+        """Returns number of characters in character matrix.
+        """
+        return self.character_matrix.shape[1]
+
+    @property
+    def root(self) -> str:
+        """Returns root of tree.
+        """
+        if self.__network is None:
+            return None
+        return [n for n in self.__network if self.__network.in_degree(n) == 0][
+            0
+        ]
+
+    @property
+    def leaves(self) -> List[str]:
+        """Returns leaves of tree.
+        """
+        if self.__network is None:
+            return None
+        return [n for n in self.__network if self.__network.out_degree(n) == 0]
+
+    @property
+    def internal_nodes(self) -> List[str]:
+        """Returns internal nodes in tree (including the root).
+        """
+        if self.__network is None:
+            return None
+        return [
+            n
+            for n in self.__network
+            if self.__network.out_degree(n) > 1
+        ]
+
+    @property
+    def nodes(self) -> List[str]:
+        """Returns all nodes in tree.
+        """
+        if self.__network is None:
+            return None
+        return [n for n in self.__network]
+
+    @property
+    def edges(self) -> List[Tuple[str, str]]:
+        """Returns all edges in the tree.
+        """
+        if self.__network is None:
+            return None
+        return [(u, v) for (u,v) in self.__network.edges]
+
+    def is_leaf(self, node: str) -> bool:
+        """Returns whether or not the node is a leaf.
+        """
+        return (node in self.leaves)
```

Comment:
How about `self.__network.out_degree(node) == 0` for O(1)?

Reply:
Good suggestion!